### DISCUSSION TITLE
Two question about ESM_MSA

### DISCUSSION THREAD

**Discussion by zedzad** [facebookresearch/esm#74] [Q&A] [answered]
Hi 
Thanks so much for sharing your great job; 
I have two question; I was wondering if you answer both:
1: After downloading the esm1b and esm-MSA, although I did not have any problem to load the esm1b, I faced the below error when I try to load the esm-MSA:
>>>in esm_msa1_t12_100M_UR50S
>>>return load_model_and_alphabet_hub("esm_msa1_t12_100M_UR50S")
>>> in load_model_and_alphabet_hub
>>>model_data = load_hub_workaround(url)
>>>in load_hub_workaround
 >>>   f"{torch.hub.get_dir()}/checkpoints/{fn}",
>>>AttributeError: module 'torch.hub' has no attribute 'get_dir'

2. Also, in the MSA-Transformer paper, are you using 1024 number of seq as input for below section? If yes, how you managed the GPU capacity to do it? In the example of Contact Prediction Examples, you only used 64 and it seems it is OK, but when increasing it to 1024, is there any trick we should use? 
>>>def read_msa(filename: str, nseq: int) -> List[Tuple[str, str]]:
>>>    """ Reads the first nseq sequences from an MSA file, automatically removes insertions."""
>>>   return [(record.description, remove_insertions(str(record.seq)))
>>>          for record in itertools.islice(SeqIO.parse(filename, "fasta"), nseq)]

Thanks so much

**Comment by joshim5** [comment, marked as answer]
Hello @zedzad, thanks for your interest in our work!
1. What version of PyTorch are you using? Please try using an updated version of PyTorch. The method `get_dir` exists in new versions of PyTorch, [for example 1.8.1,](https://pytorch.org/docs/stable/hub.html#torch.hub.get_dir) and probably earlier versions as well.
2. No, we use a max sequence length of 1024. The model was trained using 2**14 = 16384 input tokens in the MSA (that's what we could fit in memory during training), so we recommend using a similar number of tokens for inference. It's usually possible to fit more than this during inference if you have a 32GB GPU if you use `torch.no_grad()`.

Hope these pointers help! I'm closing out the issue because nothing needs to be changed in ESM, but feel free to follow up if you have more questions.

**Comment by zedzad** [comment]
Thanks so much for your prompt reply; 
Yes, I have found my pytorch was 1.5 and will update it, thanks

About the second question, what I meant was the number of sequence in the input MSA in inference mode (not the number of token); I mean when using the model only based on an MSA containing maximum 200 number of sequence as input, every thing is OK, but, when I used more than that number of sequence I get the below error: 
>>>CUDA out of memory. Tried to allocate 3.00 GiB (GPU 0; 23.65 GiB total capacity; 19.37 GiB already allocated; 2.54 GiB free; 19.99 GiB reserved in total by PyTorch)

**Reply by joshim5** [reply]
The number of sequences will depend on the length of those sequences. So, if you you have a MSA of length L (meaning every sequence in your MSA has length L), then you should use 2**14 / L sequences. This definitely fits in 32GB GPUs; should likely fit in your 23GB GPU as well. If it does not, you may need to decrease the number of sequences. Also, make sure you are using `torch.no_grad()` and that you don't have anything else unnecessary loaded on your GPU.

**Comment by zedzad** [comment]
Thanks so much for you reply 
Only, one more thing about this problem; 
In the Figure 6 in the paper, you mentioned you used 1024 sequence selected with hhfilter are used as input to MSA Transformer, So if you used this number of sequence, the number of token will exceed that threshold, no? 
Thanks again

**Reply by joshim5** [reply]
Great question. Yes, we used 1024 sequences for that experiment. This fit in memory during inference. For that experiment, we specifically used more sequences as we investigated how the model captures covariance and similarity. We wanted to ensure there were enough sequences for the model to infer coevolution even it couldn't use similarity. 
For general applications, Figure 4 shows that performance starts to cap off around 2**8 sequences.
As a practical suggestion, I'd recommend starting off with 2**14 sequences and then treating that number as a hyperparameter that you can tune on a validation set for your specific application.

**Comment by zedzad** [comment]
Oh, Great, Thanks so much for your helpful answer; much appreciated;
